src/plonk/circuit.rs

```diff
@@ -1072,11 +1156,280 @@ impl<F: Field> ConstraintSystem<F> {
+        let mut seen = vec![false; self.num_selectors];
+
+        let mut selector_map = vec![0; self.num_selectors];
+
+        for i in 0..self.num_selectors {
+            if complex_selectors[i] {
+                selector_map[i] = combinations.len();
+                // Complex selectors are on their own.
+                combinations.push(vec![i]);
+
+                // NB: We don't update `seen` here as it's a complex selector
+                // that will be ignored anyway.
+            } else {
+                // This is a simple selector.
+                if seen[i] {
+                    // This simple selector was added to a previous combination.
+                    continue;
+                }
+                selector_map[i] = combinations.len();
+                seen[i] = true; // We're definitely adding this to a combination.
+                let mut combination = vec![i];
+                assert!(degrees[i] <= max_degree); // Otherwise, self.degree() is wrong.
+
+                // Keep track of how large the substitution is.
+                let mut usage = 1;
+
+                // Keep track of how large of a gate any of the members of this
+                // combination can support.
+                let mut room = max_degree - degrees[i];
+
+                // Try to find other selectors that can join this one.
+                for j in (i + 1)..self.num_selectors {
+                    // Is there room for the new selector?
+                    if room == 0 {
+                        // Short circuit here, we've hit our limit.
+                        break;
+                    }
+
+                    // Skip complex selectors.
+                    if complex_selectors[j] {
+                        continue;
+                    }
+
+                    // Skip selectors that have been added to previous combinations
+                    if seen[j] {
+                        continue;
+                    }
+
+                    // Are the two selectors excluded from co-existing?
+                    // j > i and so (j, i) is always on the lower diagonal
+                    if exclusion_matrix[j][i] {
+                        continue;
+                    }
+
+                    // Can the new selector join the combination?
+                    if degrees[j] + usage <= max_degree {
+                        // Yes, it can. Let's do it.
+                        combination.push(j);
+                        // The remaining room available is the lesser of the
+                        // current remaining room available for previous
+                        // selector(s) after adjusting for adding this new
+                        // selector, and the room available in the new selector
+                        // after adjusting for adding the previous selectors.
+                        room = std::cmp::min(room - 1, max_degree - (degrees[j] + usage));
```

Comment:
This is correct but I think it's a complicated way of doing it. It's simpler to keep track of the maximum degree of this combination so far, and then instead of shortcutting on `room == 0`, you can shortcut on `degree_of_combination == max_degree`.

Reply:
Yeah, I switched to the way you do it in your python code.